File: core/agent.py
"""Agent基类和各种Agent实现"""
from abc import ABC, abstractmethod
from typing import Optional, Any
from .message import Message
from .llm import HelloAgentsLLM
from .config import Config


class Agent(ABC):
    """Agent基类"""
    
    def __init__(
        self,
        name: str,
        llm: HelloAgentsLLM,
        system_prompt: Optional[str] = None,
        config: Optional[Config] = None
    ):
        self.name = name
        self.llm = llm
        self.system_prompt = system_prompt
        self.config = config or Config()
        self._history: list[Message] = []
    
    @abstractmethod
    def run(self, input_text: str, **kwargs) -> str:
        """运行Agent"""
        pass
    
    def add_message(self, message: Message):
        """添加消息到历史记录"""
        self._history.append(message)
    
    def clear_history(self):
        """清空历史记录"""
        self._history.clear()
    
    def get_history(self) -> list[Message]:
        """获取历史记录"""
        return self._history.copy()
    
    def __str__(self) -> str:
        return f"Agent(name={self.name}, provider={self.llm.provider})"


class SimpleAgent(Agent):
    """简单对话Agent基类"""
    
    def run(self, input_text: str, **kwargs) -> str:
        messages = []
        if self.system_prompt:
            messages.append({"role": "system", "content": self.system_prompt})
        for msg in self._history:
            messages.append({"role": msg.role, "content": msg.content})
        messages.append({"role": "user", "content": input_text})
        response = self.llm.think(messages, **kwargs)
        self.add_message(Message(input_text, "user"))
        self.add_message(Message(response, "assistant"))
        return response


class ReActAgent(Agent):
    """ReAct (Reasoning + Acting) Agent基类"""
    def run(self, input_text: str, **kwargs) -> str:
        raise NotImplementedError("ReActAgent子类必须实现run方法")


class ReflectionAgent(Agent):
    """反思Agent基类 - 支持生成-反思-改进循环"""
    def __init__(self, name: str, llm: HelloAgentsLLM, system_prompt: Optional[str] = None,
                 config: Optional[Config] = None, max_iterations: int = 2):
        super().__init__(name, llm, system_prompt, config)
        self.max_iterations = max_iterations
    
    def run(self, input_text: str, **kwargs) -> str:
        raise NotImplementedError("ReflectionAgent子类必须实现run方法")


class PlanAndSolveAgent(Agent):
    """计划与执行Agent基类 - Plan-Solve模式"""
    def run(self, input_text: str, **kwargs) -> str:
        raise NotImplementedError("PlanAndSolveAgent子类必须实现run方法")

File: src/core/agent.ts
/**
 * Agent 基类 - 所有 Agent 的抽象基类
 */

import { Message } from "./message.js";
import { HelloAgentsLLM } from "./llm.js";
import { Config, createConfig } from "./config.js";
import { Logger, silentLogger } from "./logger.js";

/**
 * Agent 构造选项
 */
export interface AgentOptions {
  name: string;
  llm: HelloAgentsLLM;
  systemPrompt?: string;
  config?: Partial<Config>;
  /** 可选注入 Logger */
  logger?: Logger;
}

/**
 * Agent 抽象基类
 */
export abstract class Agent {
  readonly name: string;
  readonly llm: HelloAgentsLLM;
  readonly systemPrompt?: string;
  readonly config: Config;
  readonly logger: Logger;

  protected history: Message[] = [];

  constructor(options: AgentOptions) {
    this.name = options.name;
    this.llm = options.llm;
    this.systemPrompt = options.systemPrompt;
    this.config = createConfig(options.config);
    this.logger = options.logger ?? silentLogger;
  }

  /**
   * 运行 Agent（抽象方法，子类必须实现）
   */
  abstract run(
    input: string,
    options?: Record<string, unknown>
  ): Promise<string>;

  /**
   * 添加消息到历史记录
   */
  addMessage(message: Message): void {
    this.history.push(message);

    // 限制历史记录长度
    if (this.history.length > this.config.maxHistoryLength) {
      this.history = this.history.slice(-this.config.maxHistoryLength);
    }
  }

  /**
   * 清空历史记录
   */
  clearHistory(): void {
    this.history = [];
  }

  /**
   * 获取历史记录副本
   */
  getHistory(): Message[] {
    return [...this.history];
  }

  /**
   * 获取历史记录长度
   */
  getHistoryLength(): number {
    return this.history.length;
  }

  /**
   * 字符串表示
   */
  toString(): string {
    return `Agent(name=${this.name}, provider=${this.llm.provider})`;
  }
}
